drivers/gpio/gpio_keys_zephyr.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio_keys.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(zephyr_gpio_keys, LOG_LEVEL_DBG);
+
+#define DT_DRV_COMPAT zephyr_gpio_keys
+
+#define GPIO_KEYS_CFG_DEF(i)                                                                       \
+	{                                                                                          \
+		.spec = GPIO_DT_SPEC_GET(DT_DRV_INST(i), gpios),                                   \
+		.debounce_ms = DT_INST_PROP(i, debounce_interval),                                 \
+	}
+
+struct gpio_keys_data {
+	const struct device *dev;
+	gpio_callback_handler_t callback;
+	struct gpio_callback cb_data;
+	struct k_work_delayable work;
+	uint32_t debounce_ms;
+	int pin_state;
+	int is_stable;
+};
+
+static int gpio_keys_init(const struct device *dev)
+{
+	struct gpio_keys_data *data = (struct gpio_keys_data *)dev->data;
+
+	data->dev = dev;
+	data->pin_state = -1;
+	data->is_stable = 0;
+
+	return 0;
+}
+
+/**
+ * Handle debounced gpio pin state.
+ */
+static void gpio_keys_change_deferred(struct k_work *work)
+{
+	struct gpio_keys_data *data = CONTAINER_OF(work, struct gpio_keys_data, work);
+	const struct device *dev = data->dev;
+	const struct gpio_keys_config *cfg = (struct gpio_keys_config *)dev->config;
+
+	const int new_pressed = gpio_keys_get_pin_raw(dev);
+
+	LOG_DBG("gpio_change_deferred %s pin_state=%d, new_pressed=%d", dev->name, data->pin_state,
+		new_pressed);
+
+	/* If gpio hasn't changed state, nothing to do */
+	if (new_pressed == data->pin_state) {
+		data->is_stable = 1;
+		return;
+	}
+
+	data->pin_state = new_pressed;
+	data->is_stable = 1;
+
+	LOG_DBG("Calling callback %s %d", dev->name, new_pressed);
+	data->callback(dev, &data->cb_data, BIT(cfg->spec.pin));
+}
+
+static void gpio_keys_change_call_deferred(struct gpio_keys_data *data, uint32_t msec)
+{
+	int rv;
+
+	k_work_init_delayable(&data->work, gpio_keys_change_deferred);
+
+	rv = k_work_schedule(&data->work, K_MSEC(msec));
+	if (rv == 0) {
+		rv = k_work_reschedule(&data->work, K_MSEC(msec));
+	}
```

Comment:
Can't you just always call `k_work_reschedule`? Not sure I understand the logic here

Reply:
I see, the documentation seems to indicate that `k_work_reschedule` only works on non-idle work items.  But I see now that it's totally fine to use this alone. Updated code to just call `k_work_reschedule`

Reply:
Thanks, fyi I also find that bit of documentation unclear. Fancy sending an update to reword that paragraph?

Reply:
done